lang/core/os/is_virtual.go

```diff
@@ -0,0 +1,138 @@
+// modules which link with this program, contain a copy of their source code in
+// the authoritative form) containing parts covered by the terms of any other
+// license, the licensors of this program grant you additional permission to
+// convey the resulting work. Furthermore, the licensors of this program grant
+// the original author, James Shubin, additional permission to update this
+// additional permission if he deems it necessary to achieve the goals of this
+// additional permission.
+
+package coreos
+
+import (
+	"context"
+	"errors"
+	"os"
+	"runtime"
+	"strings"
+
+	"github.com/purpleidea/mgmt/lang/funcs/simple"
+	"github.com/purpleidea/mgmt/lang/types"
+)
+
+var virtualizationVendorMap = map[string]string{
+	"KVM":                    "KVM",
+	"OpenStack":              "KVM",
+	"KubeVirt":               "KVM",
+	"Amazon EC2":             "Amazon",
+	"QEMU":                   "QEMU",
+	"VMware":                 "VMware",
+	"VMW":                    "VMware",
+	"innotek GmbH":           "Oracle",
+	"VirtualBox":             "Oracle",
+	"Oracle Corporation":     "Oracle",
+	"Xen":                    "Xen",
+	"Bochs":                  "Bochs",
+	"Parallels":              "Parallels",
+	"BHYVE":                  "BHYVE",
+	"Hyper-V":                "Microsoft",
+	"Apple Virtualization":   "Apple",
+	"Google Computer Engine": "Google",
+}
+
+var dmiFilesSlice = []string{
+	"/sys/class/dmi/id/product_name",
+	"/sys/class/dmi/id/sys_vendor",
+	"/sys/class/dmi/id/board_vendor",
+	"/sys/class/dmi/id/bios_vendor",
+	"/sys/class/dmi/id/product_version",
+}
+
+func init() {
+	simple.ModuleRegister(ModuleName, "is_virtual", &simple.Scaffold{
+		T: types.NewType("func() bool"),
+		F: IsVirtual,
+	})
+}
+
+// IsVirtual is a simple function that executes three types of checks: first, we
+// check whether we're running on Linux. If that's the case, we run the two
+// different checks we have, related with the presence of virtualization and
+// containerization platforms. If any of those checks returns true, then so does
+// this function; otherwise we assume that we're not in any of those contexts
+func IsVirtual(ctx context.Context, input []types.Value) (types.Value, error) {
+	// If we implement detection for OS other than Linux, this logic will have
+	// to change
+	opersys := runtime.GOOS
+	if opersys != "linux" {
+		return nil, errors.New("we're not running on Linux, exiting")
+	}
+
+	switch {
+	case virtCheck():
+		return &types.BoolValue{V: true}, nil
+	case containerCheck():
+		return &types.BoolValue{V: true}, nil
+	default:
+		return &types.BoolValue{V: false}, nil
+	}
+}
+
+// We make use of systemd's work for detecting virtualization platforms, and
+// check if any of the keys of virtualizationVendorMap is present on any of the
+// DMI files present on dmiFilesSlice. If that's the case, then we return this
+// func as true. https://github.com/systemd/systemd/blob/main/src/basic/virt.c
+func virtCheck() bool {
```

Comment:
Don't be afraid to have this signature include an error as well, and a context as input, if that makes sense.